flang/unittests/Optimizer/InternalNamesTest.cpp

```diff
@@ -0,0 +1,179 @@
+        kinds{kinds.begin(), kinds.end()} {}
+
+  bool isObjEqual(const NameUniquer::DeconstructedName &actualObj) {
+    if ((actualObj.name == name) && (actualObj.modules == modules) &&
+        (actualObj.host == host) && (actualObj.kinds == kinds)) {
+      return true;
+    }
+    return false;
+  }
+
+private:
+  llvm::SmallVector<std::string, 2> modules;
+  llvm::Optional<std::string> host;
+  std::string name;
+  llvm::SmallVector<std::int64_t, 4> kinds;
+};
+
+void validateDeconstructedName(
+    std::pair<NameUniquer::NameKind, NameUniquer::DeconstructedName> &actual,
+    NameUniquer::NameKind &expectedNameKind,
+    struct DeconstructedName &components) {
+  EXPECT_EQ(actual.first, expectedNameKind)
+      << "Possible error: NameKind mismatch";
+  ASSERT_TRUE(components.isObjEqual(actual.second))
+      << "Possible error: DeconstructedName mismatch";
+}
+
+TEST(InternalNamesTest, doCommonBlockTest) {
+  NameUniquer obj;
+  std::string actual = obj.doCommonBlock("hello");
+  std::string actualBlank = obj.doCommonBlock("");
+  std::string expectedMangledName = "_QBhello";
+  std::string expectedMangledNameBlank = "_QB";
+  ASSERT_EQ(actual, expectedMangledName);
+  ASSERT_EQ(actualBlank, expectedMangledNameBlank);
+}
+
+TEST(InternalNamesTest, doGeneratedTest) {
+  NameUniquer obj;
+  std::string actual = obj.doGenerated("@MAIN");
+  std::string expectedMangledName = "_QQ@MAIN";
+  ASSERT_EQ(actual, expectedMangledName);
+
+  std::string actual1 = obj.doGenerated("@_ZNSt8ios_base4InitC1Ev");
+  std::string expectedMangledName1 = "_QQ@_ZNSt8ios_base4InitC1Ev";
+  ASSERT_EQ(actual1, expectedMangledName1);
+
+  std::string actual2 = obj.doGenerated("_QQ@MAIN");
+  std::string expectedMangledName2 = "_QQ_QQ@MAIN";
+  ASSERT_EQ(actual2, expectedMangledName2);
+}
+
+TEST(InternalNamesTest, doConstantTest) {
+  NameUniquer obj;
+  std::string actual = obj.doConstant({"mod1", "mod2"}, {"foo"}, "Hello");
+  std::string expectedMangledName = "_QMmod1Smod2FfooEChello";
+  ASSERT_EQ(actual, expectedMangledName);
+}
+
+TEST(InternalNamesTest, doProcedureTest) {
+  NameUniquer obj;
+  std::string actual = obj.doProcedure({"mod1", "mod2"}, {}, "HeLLo");
+  std::string expectedMangledName = "_QMmod1Smod2Phello";
+  ASSERT_EQ(actual, expectedMangledName);
+}
+
+TEST(InternalNamesTest, doTypeTest) {
+  NameUniquer obj;
+  std::string actual = obj.doType({}, {}, "mytype", {4, -1});
+  std::string expectedMangledName = "_QTmytypeK4KN1";
+  ASSERT_EQ(actual, expectedMangledName);
+}
+
+TEST(InternalNamesTest, doIntrinsicTypeDescriptorTest) {
+  NameUniquer obj;
+  std::string actual =
+      obj.doIntrinsicTypeDescriptor({}, {}, obj.IntrinsicType::REAL, {42});
+  std::string expectedMangledName = "_QCrealK42";
+  ASSERT_EQ(actual, expectedMangledName);
+
+  actual = obj.doIntrinsicTypeDescriptor({}, {}, obj.IntrinsicType::REAL, {});
+  expectedMangledName = "_QCrealK0"; // [FIXME]: Should be 4(DEFAULT KIND) ?
```

Comment:
The KindType is just an integer for this interface, so when you use `{}` as a ctor (e.g., `std::int64_t{}`) you will get the value zero.  Kinds are expected to come from the Fortran front-end. We wouldn't expect C++ to know about them. (The braces in the other calls like line 99 aren't needed, btw.)